# Table of Contents

01. Import Libraries
02. Import Data
03. Exercise 4.5
04. Task 4.5

# 01. Import Libraries

In [1]:
#import libraries - Task 4.5
import pandas as pd
import numpy as np
import os

# 02. Import Data

In [2]:
path = r'C:\Users\Cathy\2022 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
dfords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

# 03. Exercise 4.5

In [6]:
# create a data frame
df_test = pd.DataFrame()

In [7]:
# create a mixed-type column
df_test['mix'] = ['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# to check for mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# to assign data type to column
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
#check for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
#subset to see missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prods.shape

(49693, 5)

In [17]:
#create data frame without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [ ]:
#other options to remove missing values (these will overwrite dataframe):
#df_prods.dropna(subset = [‘product_name’], inplace = True)
#df_prods.dropna(inplace = True)

In [19]:
#check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
#drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

# 04. Task 4.5

In [23]:
#Step 2
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Step 2.
There is a count of 49693 rows, but the max product ID is 49688 which implies there are missing values.
A max price of 99,999 for something at a grocery store also seems extreme - possibly a filler number or an actual outlier.

In [25]:
#Step 3: check for mixed-type columns - wrangled orders (dfords)
for col in dfords.columns.tolist():
  weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (dfords[weird]) > 0:
    print (col)

Step 4: No mixed-type columns returned in step 3. This was done on the wrangledorders.csv file as instructed to import at beginning of lesson as dfords, not on the original df_ords.

In [26]:
#Step 5: check for missing values
dfords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
userr_id                       0
dtype: int64

Step 5. There are 206,209 missing values in the days_since_prior_order column. These could be users who have never made an order, one-time users, or it could be data corruption. 

In [27]:
dfords_nan = dfords[dfords['days_since_prior_order'].isnull()==True]

In [28]:
dfords_nan.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
11          11   2168274        2             1                   2   
26          26   1374495        3             1                   1   
39          39   3343014        4             1                   6   
45          45   2717275        5             1                   3   

    order_hour_of_day  days_since_prior_order  userr_id  
0                   8                     NaN         1  
11                 11                     NaN         2  
26                 14                     NaN         3  
39                 11                     NaN         4  
45                 12                     NaN         5

In [29]:
#Step 6: address missing values
dfords.shape

(3421083, 8)

In [30]:
dfords_clean = dfords[dfords['days_since_prior_order'].isnull() == False]

In [31]:
dfords_clean.shape

(3214874, 8)

The missing values were excluded by making a new data frame so analysis would not be impeded. This dataframe has 206,209 less records. If these missing values are pertinent to the analysis, a new column could be made to flag them as inactive users.

In [32]:
#Step 7: check for duplicates
dfords_dups = dfords_clean[dfords_clean.duplicated()]

In [33]:
dfords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, userr_id]
Index: []

In [ ]:
Step 7/8: no duplicates found

In [35]:
dfords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled_cleaned.csv'))

In [36]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.csv'))